### PR TITLE
minio: set version to a valid datetime string

### PR DESCRIPTION
**Pull request by jojosch**
###### Motivation for this change

Minio verifies on each RPC call that the server version returned by the API is a valid datetime string (https://github.com/minio/minio/blob/3a0e7347cad25c60b2e51ff3194588b34d9e424c/browser/app/js/web.js#L51-L53).

When returning the version as "2021-04-22T15-44-28Z" this check fails and the login (and all other RPC calls) in the minio browser are not possible:

image

Is there a cleaner way to replace just the last two dashes in "2021-04-22T15-44-28Z" with colons so that the version attribute can be left as-is and only the part in the `patchPhase` will be adjusted?

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 06kellyjac** [comment]
```
{ lib, buildGoModule, fetchFromGitHub, nixosTests }:

let
  modifyTimestamp = version:
    let
      splitTS = builtins.elemAt (builtins.split "(.*)(T.*)" version) 1;
    in
    builtins.concatStringsSep "" [ (builtins.elemAt splitTS 0) (builtins.replaceStrings [ "-" ] [ ":" ] (builtins.elemAt splitTS 1)) ];
in
buildGoModule rec {
  pname = "minio";
  version = "2021-04-22T15-44-28Z";

  src = fetchFromGitHub {
    owner = "minio";
    repo = "minio";
    rev = "RELEASE.${version}";
    sha256 = "147a4vgf2hdpbndska443axzvxx56bmc0011m3cq4ca1vm783k8q";
  };

  vendorSha256 = "0qj1zab97q8s5gy7a304wqi832y8m083cnk8hllz8lz9yjcw6q92";

  doCheck = false;

  subPackages = [ "." ];

  patchPhase = ''
    sed -i "s/Version.*/Version = \"${modifyTimestamp version}\"/g" cmd/build-constants.go
    grep "Version." cmd/build-constants.go
    sed -i "s/ReleaseTag.*/ReleaseTag = \"RELEASE.${version}\"/g" cmd/build-constants.go
    sed -i "s/CommitID.*/CommitID = \"${src.rev}\"/g" cmd/build-constants.go
  '';

  # ...
}

```

I came up with this function but it's a bit ugly lol

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 121786` at 8d9165ed run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>minio</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/patch-phase.md">patch-phase</a>

`patchPhase` should not be overridden, use `postPatch` instead.

Near pkgs/servers/minio/default.nix:25:3:
```
   |
25 |   patchPhase = ''
   |   ^
```
</li>
  </ul>
</details>

**Comment by bachp** [comment]
@ryantm will this cause issues with the automatic update if the version doesn't match the tag? If yes I prefer the solution from @06kellyjac

File: pkgs/servers/minio/default.nix [hunk 1]
```diff
@@ -1,8 +1,13 @@
 { lib, buildGoModule, fetchFromGitHub, nixosTests }:
 
+let
+  # The web client verifies, that the server version is a valid datetime string:
+  # https://github.com/minio/minio/blob/3a0e7347cad25c60b2e51ff3194588b34d9e424c/browser/app/js/web.js#L51-L53
+  versionTimestamp = "2021-04-22T15:44:28Z";
+in
```
Comment (SuperSandro2000) [inline pkgs/servers/minio/default.nix:7]:
```suggestion
```

File: pkgs/servers/minio/default.nix [hunk 2]
```diff
@@ -18,7 +23,7 @@ buildGoModule rec {
   subPackages = [ "." ];
 
   patchPhase = ''
-    sed -i "s/Version.*/Version = \"${version}\"/g" cmd/build-constants.go
+    sed -i "s/Version.*/Version = \"${versionTimestamp}\"/g" cmd/build-constants.go
```
Comment (SuperSandro2000) [inline pkgs/servers/minio/default.nix:26]:
```suggestion
    sed -i "s/Version.*/Version = \"${lib.replaceStrings [":"] ["-"] version}\"/g" cmd/build-constants.go
```

File: pkgs/servers/minio/default.nix [hunk 3]
```diff
@@ -1,8 +1,13 @@
...
 
+let
+  # The web client verifies, that the server version is a valid datetime string:
+  # https://github.com/minio/minio/blob/3a0e7347cad25c60b2e51ff3194588b34d9e424c/browser/app/js/web.js#L51-L53
+  versionTimestamp = "2021-04-22T15:44:28Z";
+in
 buildGoModule rec {
   pname = "minio";
-  version = "2021-04-22T15-44-28Z";
+  version = lib.replaceStrings [":"] ["-"] versionTimestamp;
```
Comment (SuperSandro2000) [inline pkgs/servers/minio/default.nix:10]:
```suggestion
  version = "2021-04-22T15-44-28Z";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
> @ryantm will this cause issues with the automatic update if the version doesn't match the tag? If yes I prefer the solution from @06kellyjac

Yeah it will. It would require a custom update script which is not necessary.

**Comment by ryantm** [comment]
@bachp You can add "nixpkgs-update: no auto update" to a comment in the derivation, or ask me to add it to the [Skiplist](https://github.com/ryantm/nixpkgs-update/blob/master/src/Skiplist.hs) if you don't want nixpkgs-update to touch it.

File: pkgs/servers/minio/default.nix (hunk 2 above)
Comment (jojosch) [reply pkgs/servers/minio/default.nix]: That wouldn't change the version, wouldn't it? `"2021-04-22T15-44-28Z"` -> replace `: / -` -> still `"2021-04-22T15-44-28Z"` which would be rejected by the minio browser.

**Comment by bachp** [comment]
@ryantm No I would very much like to keep auto updates, they are very handy. With the updated solution they should continue to work.

**Review by bachp — APPROVED** [review]

File: pkgs/servers/minio/default.nix [hunk 4]
```diff
@@ -1,5 +1,14 @@
 { lib, buildGoModule, fetchFromGitHub, nixosTests }:
 
+let
+  # The web client verifies, that the server version is a valid datetime string:
+  # https://github.com/minio/minio/blob/3a0e7347cad25c60b2e51ff3194588b34d9e424c/browser/app/js/web.js#L51-L53
+  versionToTimestamp = version:
+    let
+      splitTS = builtins.elemAt (builtins.split "(.*)(T.*)" version) 1;
+    in
+    builtins.concatStringsSep "" [ (builtins.elemAt splitTS 0) (builtins.replaceStrings [ "-" ] [ ":" ] (builtins.elemAt splitTS 1)) ];
```
Comment (SuperSandro2000) [inline pkgs/servers/minio/default.nix:10]: Please add a comment with an example like:

```
2021-04-22T15-44-28Z -> ?
```

**Review by bachp — APPROVED** [review]

**Pull request merged**